### ISSUE TITLE
[Bug] The Qwen3 Jinja2 chat template supports switching between “thinking” and “non-thinking” modes. In the current code, this leads to the draft model’s training objective being inconsistent with the target model’s expectations.

### ISSUE THREAD

**Issue by Libertax-coder** [sgl-project/SpecForge#296]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/SpecForge/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

I noticed that when using the generate_data_by_target.py script to generate data, the data is stored in standard ChatML (I modified the code to ensure that Qwen3 uses non-thinking mode during data generation). When using the build_eagle3_dataset_cache.py script for tokenization, parse.py uses Qwen3’s Jinja2 chat template to package the ChatML data. The source code is:

conversation = self.tokenizer.apply_chat_template(
                messages, tokenize=False, add_generation_prompt=False, **kwargs
            )

The input message is:
<img width="2170" height="132" alt="Image" src="https://github.com/user-attachments/assets/bb2cef40-8ad6-4298-941f-6622bd5d2a78" />
The output conversation is:

<img width="662" height="340" alt="Image" src="https://github.com/user-attachments/assets/758eb166-81cc-448b-95e5-e4118f026583" />

<img width="438" height="340" alt="Image" src="https://github.com/user-attachments/assets/0c992caf-7bcf-4cab-9c46-a90225fc596f" />

You’ll notice that an empty thinking block appears in the last assistant message, which is due to the logic in the Jinja2 template. In this case, after applying the loss mask, the result is as follows:

<img width="938" height="556" alt="Image" src="https://github.com/user-attachments/assets/c2e0ce7c-7c87-4ce2-8cb5-0614c5fa74d4" />

<img width="1154" height="642" alt="Image" src="https://github.com/user-attachments/assets/6f9a1422-78d7-47e8-a7f6-3e600405e1af" />

This training data is telling the draft model that the red part is the input and the green part is the target model's output, but this is not actually the case. When Qwen3 series dense models receive user input ending with <im_start>\nassistant\n, they automatically trigger the thinking mode, and the model actually tends to output <think> to engage in deep thinking, whereas the green part (trainable tokens) in the training data does not contain thinking content. In other words, the tokens used to train the draft model are not the tokens that the target model expects to output—this is a mismatch.
This is because Qwen3's Jinja2 template logic is designed for thinking models in multi-turn conversations. Its logic is to clear the thinking content from previous conversation rounds during multi-turn dialogue. Using this template directly in training scripts leads to the aforementioned mismatch issues, causing errors in EAGLE3 training.
Despite this, the EAGLE3 model trained in this way still has some usability, because coincidentally, the last turn of the multi-turn conversation data happens to be reasonable training data that meets expectations (all non-final turns are mismatched training data).

I’m not sure I explained it clearly—this is a subtle detail. As far as I know, most thinking models clear the previous “thinking” content when applying the chat template, which is detrimental to training the Eagle3 model.

### Reproduction

generate_data_by_target.py
build_eagle3_dataset_cache.py

### Environment

This is unrelated to the environment

**Comment by jiapingW** [comment]
Maybe the pr [here](https://github.com/sgl-project/SpecForge/pull/258) can solve your problem.

**Comment by Libertax-coder** [comment]
In fact, I don’t believe that removing the empty thinking block <think>\n\n</think>\n\n in [here](https://github.com/sgl-project/SpecForge/pull/258) would resolve the training issue when `enable_thinking=False`.

When Qwen3 receives input ending with `<|im_start|>assistant\n`, it tends by default to output `<think>thinking content</think>`.

To disable thinking, we make the input end with `<|im_start|>assistant\n<think>\n\n</think>\n\n`, after which Qwen3 will tend to skip the thinking process.

```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
Hello! How can I assist you today? 😊<|im_end|>
```

When we feed such a text sequence to the target model (Qwen3 thinking series models), after `<|im_start|>assistant\n`, the target model is inclined to start thinking, while the draft model receives the supervision signal "Hello! How can I assist you today? 😊<|im_end|>" (without thinking).

This creates a mismatch.

In actual multi-turn conversations, when thinking is disabled, Qwen3 processes data as follows:

**Input:**
```json
[{"role": "user", "content": "greetings"}]
```

**After Qwen3's chat template processing:**
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
<think>

</think>

```

This is the actual input string, after which Qwen3 responds:
```
Hello! How can I assist you today? 😊<|im_end|>
```

If the user continues with a question like "What's the weather like today in New York?", Qwen3's chat template **removes the thinking block from the last assistant message before the new user question**, so:

**After Qwen3's chat template processing:**
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
Hello! How can I assist you today? 😊<|im_end|>
<|im_start|>user
What's the weather like today in New York?<|im_end|>
<|im_start|>assistant
<think>

</think>

```

This is the actual input string, after which Qwen3 responds, e.g., "It's sunny and around 25°C in New York today. A great day to be outside!"

**This means that when training the draft model in non-thinking mode, we can only feed samples like this:**

**Sample 1:**
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
<think>

</think>

Hello! How can I assist you today? 😊<|im_end|>
```

**Sample 2:**
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
Hello! How can I assist you today? 😊<|im_end|>
<|im_start|>user
What's the weather like today in New York?<|im_end|>
<|im_start|>assistant
<think>

</think>

It's sunny and around 25°C in New York today. A great day to be outside!<|im_end|>
```

**And only the tokens after `<|im_start|>assistant\n<think>\n\n</think>\n\n` are used for loss calculation.**

**If we calculate the loss for "Hello! How can I assist you today. 😊" in Sample 2, the problem I mentioned above occurs:**

The target model (Qwen3) expects to think after seeing `<|im_start|>assistant\n`, but the draft model receives supervision signal without thinking. This creates a **distribution mismatch** between the draft model and target model.

Additionally, if we attempt to construct input samples in the form of:

Adding empty thinking blocks to all assistant turns:
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
<think>

</think>

Hello! How can I assist you today? 😊<|im_end|> ← Keep empty thinking block
<|im_start|>user
What's the weather like today in New York?<|im_end|>
<|im_start|>assistant
<think>

</think>

It's sunny and around 25°C in New York today.<|im_end|>
```

And hoping to calculate loss for model outputs after both `<think>\n\n</think>\n\n` blocks, this is equally unreasonable.

Because in reality, when Qwen3 (target model) generates `It's sunny and around 25°C in New York today.<|im_end|>`, its prefix is:
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
Hello! How can I assist you today? 😊<|im_end|>
<|im_start|>user
What's the weather like today in New York?<|im_end|>
<|im_start|>assistant
<think>

</think>
```

If we train this way, we would be telling the draft model that when generating `It's sunny and around 25°C in New York today.<|im_end|>`, the prefix is:
```
<|im_start|>user
greetings<|im_end|>
<|im_start|>assistant
<think>

</think>

Hello! How can I assist you today? 😊<|im_end|> ← Keep empty thinking block
<|im_start|>user
What's the weather like today in New York?<|im_end|>
<|im_start|>assistant
<think>

</think>
```

This also causes a **prefix mismatch between the target model and draft model**.

**Comment by jiapingW** [comment]
I understand what you mean. It's because Qwen's own tokenizer template doesn't support the final role in a multi-turn dialogue being the assistant, but it expects the user. It tries to ensure the final input aligns with both "think" and non-"think" modes by adding a generated template. This issue can be avoided by concatenating prompts and training only the final dialogue, but this approach isn't compatible with other models. You can refer to an earlier code implementation: https://github.com/sgl-project/SpecForge/pull/182. We are also testing it with the code below. 

```python
from transformers import AutoTokenizer
from specforge.data.preprocessing import preprocess_conversations
from specforge.data.template import TEMPLATE_REGISTRY
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-30B-A3B", trust_remote_code=True)
messages = [[ 
{"role": "user", "content": "Who are you?"}, 
{"role": "assistant", "content": "I am a model."}, 
{"role": "user", "content": "What is your name?"}, 
{"role": "assistant", "content": "My name is QwQ."}, 
{"role": "user", "content": "What is 1+1"}, 
{"role": "assistant", "content": "=2."}
]]
chat_template = TEMPLATE_REGISTRY.get('qwen3-thinking')
max_length = 2048

conversation = tokenizer.apply_chat_template( 
messages, 
tokenize=False, 
add_generation_prompt=False, 
enable_thinking=True, 
)

print(conversation)

conversation = preprocess_conversations(tokenizer, messages, chat_template, max_length)
print([tokenizer.decode(conversation['input_ids'][0].view(-1))])
```
